### PR TITLE
Ceiling Diffuser Convection Model Causes Oscillating Winter Design Day Loads

### LINKED ISSUE #8091: Oscillating Winter Designday Zone Load with Ceiling Diffuser Model
Issue overview
--------------
User file specifies Ceiling Diffuser model as interior convection algorithm. Zone Ideal load for winter designday shows oscillating behavior rather than the expected steady state. This is checked with multiple versions of EnergyPlus and the behavior started in Version 9.2. The following two changes are suspected to be related:
[https://github.com/NREL/EnergyPlus/pull/7300](https://github.com/NREL/EnergyPlus/pull/7300)
[https://github.com/NREL/EnergyPlus/pull/7444](https://github.com/NREL/EnergyPlus/pull/7444)

### Details
Some additional details for this issue (if relevant):
 - Platform Windows 10 64bit
 - Version of EnergyPlus 9.2.0, 9.3.0
 - Helpdesk ticket number 15486

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added 
[Comparison_CeiingDiffuserModel.xlsx](https://github.com/TiejunWu/EnergyPlus-IDF-Files/blob/master/Comparison_CeiingDiffuserModel.xlsx)
[LoadCalculation_9.2.idf](https://github.com/TiejunWu/EnergyPlus-IDF-Files/blob/master/LoadCalculation_9.2.idf)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by RKStrand**
Pull request overview
---------------------
 - Fixes #8091 
 - Recent fixes to the ceiling diffuser model to eliminate a discontinuity resulted in an unusual situation where the solution for a particular input file was resulting in oscillating heating loads for a design day that should have resulted in a stable, constant heating load profile.  This oscillation started occurring when a previous fix was entered into the code.  When stepping through things in the debugger, it appears that the rate of change for the convection coefficient was such that the solution was never achieved because the changes in the convection coefficient caused the solution to "overshoot" the actual solution.  The problem was that large percentage changes in the convection coefficient were possible even with the switch to a transition zone between purely natural and purely forced convection.  Even though the slope of the convection coefficient curve was larger elsewhere, the changes seen in the debugger for low ACH values near the natural convection limit where around 40% as the algorithm tried in vain to find a stable solution.  This was fixed by increasing the range of the mixed mode convection section.  Previously, this was set to 0.5-1.0ACH.  This has now been set to 0.5-3.0ACH.  The reason for choosing 3.0ACH as the upper bound for mixed mode convection is that the correlation for the ceiling diffuser was based on data down to 3.0ACH.  For mixed mode convection, a straight linear interpolation between the convection coefficient at 3.0ACH and the natural convection value (assumed to be the value at 0.5ACH) is used.  The IDF that resulted in #8091 now results in a stable, constant heating load as expected.

One test case results in some changes to the output--_SmallOffice_Dulles.  This file uses the ceiling diffuser method for calculating interior convection so it is expected that this file would see some differences.  The changes to the output are relatively minimal except in Joules (which is a pretty small unit).  So, it seems that this differences are expected and reasonable.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [upon request] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [see above] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: doc/engineering-reference/src/surface-heat-balance-manager-processes/inside-heat-balance.tex [hunk 1]
```diff
@@ -884,7 +884,7 @@ \subsubsection{Ceiling Diffuser Algorithm}\label{ceiling-diffuser-algorithm}
...
+It should be noted that the conditions under which the correlations were developed included ACH values down to around 3.0.  So, within EnergyPlus, for ACH values of 3.0 or greater, the ceiling diffuser correlations as shown above are used.  Below ACH values of 0.5, the natural convection is determined using the TARP Algorithm (see above).  Between ACH values of 0.5 and 3.0, convection is assumed to be a mix between natural and forced convection.  Thus, between these two ACH values, EnergyPlus will use a linear interpolation between the natural convection algorithm value and the forced convection value at ACH=3.0 to avoid any discontinuities in the convection coefficient calculated and also maintain stability within the solution.
```
Comment (mitchute) [inline doc/engineering-reference/src/surface-heat-balance-manager-processes/inside-heat-balance.tex:887]: Good summary.

File: tst/EnergyPlus/unit/ConvectionCoefficients.unit.cc [hunk 2]
```diff
@@ -870,6 +870,75 @@ TEST_F(ConvectionCoefficientsFixture, TestCalcFisherPedersenCeilDiffuserNatConv)
 
 }
 
+TEST_F(ConvectionCoefficientsFixture, TestCalcFisherPedersenCeilDiffuserCorrelations)
```
Comment (mitchute) [inline tst/EnergyPlus/unit/ConvectionCoefficients.unit.cc:873]: Tests looks good.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 3]
```diff
@@ -6397,7 +6397,7 @@ namespace ConvectionCoefficients {
         if (ACH <= 0.5) {
             return Hnatural;
         } else {
-            return Hnatural + ((Hforced - Hnatural) * ((ACH - 0.5) / 0.5));
+            return Hnatural + ((Hforced - Hnatural) * ((ACH - 0.5) / 2.5)); // range for interpolation goes from ACH=0.5 to ACH=3.0 or a range of 2.5
```
Comment (mitchute) [inline src/EnergyPlus/ConvectionCoefficients.cc:6400]: The equation looks good and the comment is a nice touch to help remind us later.

**Review by mitchute — APPROVED** [review]
There was a ton of work done over on the original issue to track this down, but these changes are straightforward. I'll build tests now and merge if all is good.

**Comment by mitchute** [comment]
Tests look good to me. I ran ```_SmallOffice_Dulles``` with the Chicago and Newark weather files and am not seeing the newly added (per the last CI run) Severe CheckWarmupConvergence errors. The tolerance on these zone loads is probably too tight anyway. Regardless, they didn't show up in my local testing. Merging.

**Comment by mjwitte** [comment]
> Tests look good to me. I ran _SmallOffice_Dulles with the Chicago and Newark weather files and am not seeing the newly added (per the last CI run) Severe CheckWarmupConvergence errors. The tolerance on these zone loads is probably too tight anyway. Regardless, they didn't show up in my local testing.

Hmmm, that's probably worth a followup issue. That's one of only 3 test files that use this convection option. Try again running locally with design days simulation and you should be able to reproduce the error.

**Comment by mitchute** [comment]
Thanks, @mjwitte. I took another look and now I'm confused. 

CI shows that the warnings were added. See here: http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2020-08/6729dbe590f6808441d17b9ac36ba9085de91c06-aa2bc4186f4d77c5ee7436e206e4396aa726bbfc/_SmallOffice_Dulles/x86_64-Linux-Ubuntu-18.04-gcc-7.5/eplusout.err.diff.html

But locally I'm seeing the warnings as *being present on develop* (at 8f2a3d8b3 before this was merged), but *not* on the branch (at aa2bc4186). This is what we want to see, but I'm confused as to why CI shows it the other way around.

**Comment by RKStrand** [comment]
So, I'm seeing what @mitchute is saying with regards to the CI results.  Is it "possible" that if develop was much more advanced because a lot of pre-release code changes that it was considered the "+"?  That would see backwards to me, but is it possible?

What I would have anticipated is what you both saw--the fix eliminates those convergence warnings because it should get to the solution more gracefully.  I would have guessed that develop would have a harder time settling down specifically because it was unstable as shown by the results which oscillated.  And that's backed up by what you both saw running this locally.  I'm not sure why CI has it differently.

What else needs to happen here?  I realize that it has already been merged, but I want to make sure this is really resolved and "done".

**Comment by mjwitte** [comment]
@RKStrand I would pull current develop (which has this merged) and run _SmallOffice_Dulles both as-is and with design days only and see if you get any warmup convergence errors. If not, then, never mind.

**Comment by mitchute** [comment]
I just ran current develop as you suggested, @mjwitte. No convergence errors in either case.

**Comment by mjwitte** [comment]
Ok, then, we'll just blame @Myoldmopar for any CI reporting glitches.

**Comment by RKStrand** [comment]
Thanks, @mitchute!  You beat me to it.  I'm glad to see that there are no convergence errors in develop.  I'll give CI a break--I think it got confused.  Probably stress related.  I'll just assume that this is all good now.  Thanks everyone for your contributions to this defect/PR!

**Comment by mjwitte** [comment]
And for good measure, running v9.4.0-IOFreeze with _SmallOffice_Dulles for design days only does generate warmup convergence errors. So, we're done here.

**Comment by RKStrand** [comment]
👍😁🙌🏻🥳

**Comment by Myoldmopar** [comment]
> Ok, then, we'll just blame @Myoldmopar for any CI reporting glitches.

bitmoji

**Pull request merged**